phd-tests/framework/src/test_vm/config.rs

```diff
@@ -119,14 +123,29 @@ impl<'dr> VmConfig<'dr> {
         self
     }
 
+    pub fn boot_order(&mut self, disks: Vec<&'dr str>) -> &mut Self {
+        self.boot_order = Some(disks);
+        self
+    }
+
     pub fn boot_disk(
         &mut self,
         artifact: &'dr str,
         interface: DiskInterface,
         backend: DiskBackend,
         pci_device_num: u8,
     ) -> &mut Self {
-        self.boot_disk = DiskRequest {
+        // If you're calling `boot_disk` directly, you're probably not specifying an explicit boot
+        // order. the _implicit_ boot order seems to be the order disks are created in, so preserve
+        // that implied behavior by having a boot disk inserted up front.
+        //
+        // XXX: i thought that the implicit order would be *pci device order*, not disk creation
+        // order?
+        // XXX: figure out what to do if there is both an explicit boot order and a call to
+        // `.boot_disk`.
+
+        self.disks[0] = DiskRequest {
+            name: "boot-disk",
             interface,
             backend,
             source: DiskSource::Artifact(artifact),
```

Comment:
> the framework could also do something else by default, e.g. mark the first attached disk as the boot disk and allow tests to clear that designation or set it to something else. This might actually be a little more natural for test authors, now that I think about it; one of my design principles for PHD is to make tests as easy to write as possible, so we should do whatever we think will provide the least friction for test authors here.

i think this is effectively in line with the status quo as well: `.boot_disk` is _probably_ going to be the first PCI device, so it ends up the boot disk in all the tests anyway. it'll also match what i expect to end up being the control plane's configuration. writing tests for multi-device `fwcfg`-based boot orders has been.. informative.

Reply:
took this mostly along the lines of your third-and-a-half point: https://github.com/oxidecomputer/propolis/pull/756/commits/617dd0c870ce382f3208d1b5f42a6eb1c94df5bb - "boot_device()" replaces an existing disk by the same name, if one is there, and puts the new disk at the front of a boot order, creating one if there isn't one. this means that all tests by default have a boot order which will have at least `boot-disk` at the front.

`fn boot_order()` still exists to override this for the `boot_order.rs` tests that need to set it more specifically, and `fn clear_boot_order()` for cases where we really didn't want the default behavior. but i think this makes things better for the median test and gives us the right tools when we need them.